Implement EvalQueryEngineTool

d-mariano opened the pull request:
# Description

### Notice
I would like input on this PR from the `llama-index` team. **If the team agrees with the need and approach, I will provide unit tests, documentation updates, and Google Colab notebooks as required.**

### Summary
The reason for this change is to provide a plug-and-play method of using on-the-fly evaluation for tools within an agent. Some areas I would like feedback on:
1. Handling this on the tool-level makes sense for the use-case
2. This is a good approach for addressing the use-case, and aligns with the `llama-index` vision
3. Anything else you can think of! <3

If the team decides this is a good approach, I will provide unit tests, documentation updates, and Google Colab notebooks as required.  

Implements the `EvalQueryEngineTool`:
* Inherits from `QueryEngineTool`
* Uses a given evaluator to evaluate query engine responses
* If the response fails evaluation:
  * The tool output is manipulated to deter the LLM from using it
  * The evaluator feedback is used as a reason for failure

## Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

logan-markewich commented on llama-index-core/llama_index/core/tools/query_engine.py:
```diff
@@ -61,18 +61,7 @@ def metadata(self) -> ToolMetadata:
         return self._metadata
 
     def call(self, *args: Any, **kwargs: Any) -> ToolOutput:
-        if args is not None and len(args) > 0:
```
I think these checks are still needed?

The LLM will often not write entirely correct inputs for a query engine tool

`{"query": "..."}`, `{"input": "..."}`, and so on.

A query engine only has one input (the query string) so we need to kind of validate that behaviour a bit

I suggest doing a full method override rather than relying on `super()` in the new tool you created

logan-markewich commented:
@d-mariano Yea I think I like this approach. If this was combined with an agent that had the chance to "retry" tool calls, it would be pretty powerful (and also helps avoid tracebacks when calling a tool)

d-mariano commented:
> @d-mariano Yea I think I like this approach. If this was combined with an agent that had the chance to "retry" tool calls, it would be pretty powerful (and also helps avoid tracebacks when calling a tool)

@logan-markewich great! And totally agree, that's the idea down the road. I would like to test this with the simple ReAct implementation. Then, after that, test it with more complex use-cases like within a query pipeline.

Also curious what implications this has on a QueryPlan, but that doesn't need to be addressed today.


Okay, great. So with this, I can implement some UTs, of course...should I introduce this in the docs with a notebook example as well? <3

logan-markewich commented:
@d-mariano yea may as well make this PR a little more complete 💪

d-mariano commented:
@logan-markewich I've made a few updates so far:
1. I corrected the change to update `ToolOutput.content` rather than `ToolOutput.raw_output`, the agents use `content` and the entire point of this change is to inform agents on evaluation failures
2. I've added a test suite

I would like to update the `llama-index` documentation with a section for `EvalQueryEngine`. Here I can write up a blurb on the purpose of it and even include a notebook example of it being used with an agent. Are these docs included in this repo? (I haven't looked around, I can also do that at some point.)

I need to get at some other stuff so I'll have this open a bit longer.

logan-markewich commented:
@d-mariano yea, there's a `docs` folder in this repo 💪 it's all markdown.

There should be some existing evaluation sections you can maybe hook into? Or maybe the section on tools would be better

logan-markewich commented on docs/examples/tools/eval_query_engine_tool.ipynb:
This makes sense! You can also garuntee that it calls the wrong tool first by using `.chat(..., tool_choice="<tool_name>")`

d-mariano replied:
Oh yeah that's wayyyyy better. I'm going to do that.

d-mariano replied:
@logan-markewich some updates:

1. Using `tool_choice` didn't work, the LLM chose the tool it thought it should use anyway...it even used the tool name for context (I didn't mention Uber financials in the description, but "uber" was still in the tool name)
2. I cannot use the [AnswerRelevancyEvaluator](https://docs.llamaindex.ai/en/stable/examples/evaluation/answer_and_context_relevancy.html#evaluating-answer-and-context-relevancy-separately) because it does not seem to set `passing` and it also kept producing invalid results when testing w/ Gemini

So with that...I've fallen back to using prompt engineering to force the model to choose a specific tool first. I also fell back to using a simple [RelevancyEvaluator](https://docs.llamaindex.ai/en/stable/examples/evaluation/relevancy_eval.html)

Let me know what you think here, you can see the outputs working. I also noticed that `HuggingFaceEmbedding` seemed to fail when using `0.10.18` due to something related to `ServiceContext`. I can post an issue if one already hasn't been posted.

logan-markewich replied:
I feel like tool choice should have worked. I'll try giving the notebook a run.

I'm not aware of any issues with huggingface embedding, it works fine for me...

d-mariano replied:
Curious to see what happens on your end! For extra context, the tool names were `lyft_10k` and `uber_10k`. You may also want to remove the instruction to choose one of them first and see if using `tool_choice` works.

RE: huggingface, I'll see if I can't reproduce it.

logan-markewich replied:
@d-mariano yea I fixed this, some of your new code introduced some circular imports I think

I also tried the notebook. I didn't realize you were using a react agent (so I switched it to an openai agent), so now the tool choice thing works.

Although, seeing some odd behaviour in the eval (see the last agent call)

d-mariano replied:
@logan-markewich thanks for taking a look.

**RE: circular imports**
I'm sorry I should have caught this. I am a bit confused as to why what was happening [here](https://github.com/run-llama/llama_index/pull/11679/commits/f99a927fadf74cb1d17b56346ff8ce06dfab391c#diff-2e296fd709df92efdd63b1a1a21eb592a6509c2b992a5f8d83133a224aad0f67) would cause circular import, though.

In fact I just realized that in [core/evaluation/__init__.py](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/evaluation/__init__.py#L55-L58), `AnswerRelevancyEvaluator`, `BaseEvaluation`, and `EvaluationResult` are all imported.

**RE: Agent**
1. Ahh okay, so `tool_choice` really can't be used in any other agent? Good to know.
2. Regarding the output, that's quite odd. Somehow I still have hit my OpenAI quota so I can't test this...but I would be curious to know what the reason would be for failing evaluation. We could try using the `AnswerRelevancyEvaluator` evaluator here just to get a reason.
3. Did we actually need to reset the agent? Or was this to ensure that there was no contamination between test cases as more of a precaution?

logan-markewich replied:
3. That was just mostly me testing if the chat history was changing the tool calling and/or eval lol

logan-markewich replied:
(Also yes, feel free to fix my spaghetti code fix haha)

d-mariano replied:
@logan-markewich I think I got it! I increased the `similarity_top_k` for the query engines from 3 to 5. I guess this is what I must do when taking the local embedding model route.

logan-markewich replied:
Nice!

d-mariano replied:
@logan-markewich let me know if there is anything else blocking this.
